Add NullableStrategy

eugene-borovov opened the pull request:
<!--
Fill in the relevant information below to help triage your issue.

Pick the target branch based on the following criteria:
  * Documentation improvement: master branch
  * Bugfix: master branch
  * QA improvement (additional tests, CS fixes, etc.) that does not change code
    behavior: master branch
  * New feature, or refactor of existing code: develop branch

You MUST provide a signoff in your commits for us to be able to accept your
patch; you can do this by providing either the --signoff or -s flag when using
"git commit". Please see the project contributing guide and
https://developercertificate.org for details.
-->

|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | yes
| RFC           | no
| QA            | no

### Description
This fix adds support null values for all strategies without any BC break. This is decorator with pre-check for null value.
I assume this is a useful addition.
<!--
Tell us about why this change is necessary:
- Are you fixing a bug or providing a failing unit test to demonstrate a bug?
  - How do you reproduce it?
  - What did you expect to happen?
  - What actually happened?
  - TARGET THE master BRANCH

- Are you adding documentation?
  - TARGET THE master BRANCH

- Are you providing a QA improvement (additional tests, CS fixes, etc.) that
  does not change behavior?
  - Explain why the changes are necessary
  - TARGET THE master BRANCH

- Are you fixing a BC Break?
  - How do you reproduce it?
  - What was the previous behavior?
  - What is the current behavior?
  - TARGET THE master BRANCH

- Are you adding something the library currently does not support?
  - Why should it be added?
  - What will it enable?
  - How will the code be used?
  - TARGET THE develop BRANCH

- Are you refactoring code?
  - Why do you feel the refactor is necessary?
  - What types of refactoring are you doing?
  - TARGET THE develop BRANCH
-->

eugene-borovov committed:
Add NullableStrategy

Signed-off-by: Eugene Borovov <eugene.borovov@gmail.com>

eugene-borovov committed:
Make Psalm happy

Signed-off-by: Eugene Borovov <eugene.borovov@gmail.com>

eugene-borovov commented:
Pull request closes #42
This fix can be used as a workaround for #37

froschdesign commented on src/Strategy/NullableStrategy.php:
```diff
@@ -0,0 +1,55 @@
+     */
+    private $strategy;
+
+    /**
+     * @param StrategyInterface $strategy
+     */
```
Can be remove the because the same information is already included in the method signature.

froschdesign commented on test/Strategy/NullableStrategyTest.php:
```diff
@@ -0,0 +1,75 @@
+use Laminas\Hydrator\Strategy\StrategyInterface;
+use PHPUnit\Framework\TestCase;
+
+/**
+ * @covers \Laminas\Hydrator\Strategy\NullableStrategy
+ */
```
Can be removed without problems.

eugene-borovov replied:
Removed

froschdesign commented on test/Strategy/NullableStrategyTest.php:
```diff
@@ -0,0 +1,75 @@
+ */
+class NullableStrategyTest extends TestCase
+{
+    /**
+     * @var bool
+     */
+    protected $backupStaticAttributes = false;
+
+    /**
+     * @var bool
+     */
+    protected $runTestInSeparateProcess = false;
```
Are these options needed for this test class?

froschdesign commented on test/Strategy/NullableStrategyTest.php:
```diff
@@ -0,0 +1,75 @@
+        self::assertEquals('hydrated value', $nullableStrategy->hydrate('original value'));
+    }
+
+    public function testHydrateNullValue() : void
```
Only the value `null` is tested here but the strategy also handles an empty string.

froschdesign commented on test/Strategy/NullableStrategyTest.php:
```diff
@@ -0,0 +1,75 @@
+        self::assertEquals('hydrated value', $nullableStrategy->hydrate('original value'));
+    }
+
+    public function testHydrateNullValue() : void
+    {
+        $strategy = $this->createMock(StrategyInterface::class);
+        $strategy->expects(self::never())
+            ->method('hydrate');
+        $nullableStrategy = new NullableStrategy($strategy);
+
+        self::assertNull($nullableStrategy->hydrate(null));
+    }
```
Only the value `null` is tested here but the strategy also handles an empty string.

froschdesign commented on src/Strategy/NullableStrategy.php:
```diff
@@ -0,0 +1,55 @@
+        return $this->strategy->extract($value, $object);
+    }
+
+    /**
+     * Check the given value for NULL so that it can be hydrated by the hydrator.
+     *
+     * {@inheritDoc}
+     */
+    public function hydrate($value, ?array $data = null)
+    {
+        if ($value === '' || $value === null) {
+            return null;
+        }
+
+        return $this->strategy->hydrate($value, $data);
+    }
```
The description of this method and also the name of this strategy class does not contains some hints that an empty string is also handled.

froschdesign commented on test/Strategy/NullableStrategyTest.php:
```diff
@@ -0,0 +1,75 @@
+ */
+class NullableStrategyTest extends TestCase
+{
+    /**
+     * @var bool
+     */
+    protected $backupStaticAttributes = false;
+
+    /**
+     * @var bool
+     */
+    protected $runTestInSeparateProcess = false;
```
Are these options needed for this test class?

eugene-borovov replied:
Without these options Psalm check could not be passed. May be Psalm need some tweeks.

eugene-borovov replied:
I removed these options. I dont understand why Psalm complains about my tests.

froschdesign reviewed (changes requested):
Thanks for your contribution! 👍 

I added some comments with some hints.

eugene-borovov committed:
Added ability treat empty value as null

Signed-off-by: Eugene Borovov <eugene.borovov@gmail.com>

eugene-borovov commented:
I realized that treating an empty string as null should be configurable. Therefore, I added an option in the constructor that is disabled by default.

weierophinney committed:
qa: fix errors flagged by psalm

- Adds phpunit plugin for psalm
- Fixes some errors flagged by re-running psalm --shepherd
- Updates the psalm baseline, as several are expected and uncorrectible

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
docs: document NullableStrategy usage

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>
